File: app/src/main/java/com/example/finances/purchases/BillingClientHelper.java
package com.example.finances.purchases;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.android.billingclient.api.SkuDetailsResponseListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class BillingClientHelper {

    public final static String SUBSCRIPTION_MONTH = "rsstudy.month";

    private Context context; //Контекст приложения
    private BillingClient billingClient; //Платежный клиент
    private List<Purchase> purchasesSub; //Список подписок пользователя

    //Конструктор для помощника
    public BillingClientHelper (Context context){
        this.context = context; //Присваиваем контекст
        purchasesSub = new ArrayList<>(); //Инициализация списка
        create(); //Создаем клиент
        startConnection(); //Запускаем соединение
    }

    //Функция создания нового платежного клиента
    private void create(){
        //Создаем новый платежный клиент
        billingClient = BillingClient.newBuilder(context).
                setListener((billingResult, list) ->
                    getPurchases()).
                enablePendingPurchases().
                build();
    }

    //Функция подключения к Google Play
    private void startConnection(){

        //Запускаем соединение с Google Play
        billingClient.startConnection(new BillingClientStateListener() {

            //Соединение не установлено
            @Override
            public void onBillingServiceDisconnected() {
                billingClient.startConnection(this); //пробуем еще раз подключиться
            }

            //Соединение установлено
            @Override
            public void onBillingSetupFinished(@NonNull @NotNull BillingResult billingResult) {
                if (isOK(billingResult)) {

                    //Проверяем, поддерживает ли устройство пользователя подписки
                    if (isOK(billingClient.isFeatureSupported(BillingClient.FeatureType.SUBSCRIPTIONS)))
                        getPurchases(); //Получаем список подписок пользоваетеля
                }
            }
        });
    }

    //Функция проверки на ответ OK
    private boolean isOK(BillingResult billingResult){
        return billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK;
    }

    //Функция для получения всех покупок пользователя
    private void getPurchases(){
        billingClient.queryPurchasesAsync(BillingClient.SkuType.SUBS, (queryBillingResult, list) -> {
            if(isOK(queryBillingResult)) {
                List<Purchase> local = list;
                purchasesSub.addAll(local);
            }
        });
    }

    //Функция проверки наличия подписки у пользователя
    public boolean isSub(String subscriptionName){
        return purchasesSub.stream().filter(p -> p.getSkus().contains(subscriptionName)).findAny().orElse(null) != null;
    }

    public void subscribe(Activity activity, String subscriptionName){

        List<String> skuList = new ArrayList<> ();
        skuList.add(subscriptionName);

        SkuDetailsParams.Builder params = SkuDetailsParams.newBuilder();
        params.setSkusList(skuList).setType(BillingClient.SkuType.SUBS);
        billingClient.querySkuDetailsAsync(params.build(),
                (billingResult, list) -> {
                    Log.e("SIZE", String.valueOf(list.size()));
                    if(isOK(billingResult)) {
                        SkuDetails skuDetails = list.get(0);

                        BillingFlowParams billingFlowParams = BillingFlowParams.newBuilder()
                                .setSkuDetails(skuDetails)
                                .build();

                        billingClient.launchBillingFlow(activity, billingFlowParams);
                    }
                });
    }

    //Функция завершения работы с клиентом
    public void finish(){
        billingClient.endConnection();
    }
}

File: app/src/main/java/com/example/finances/purchases/BillingClientHelper.kt
import android.app.Activity
import android.content.Context
import com.android.billingclient.api.*

class BillingClientHelper(context: Context) : PurchasesUpdatedListener {

    interface OnQueryProductsListener {
        fun onSuccess(products: List<SkuDetails>)
        fun onFailure(error: Error)
    }

    interface OnPurchaseListener {
        fun onPurchaseSuccess(purchase: Purchase?)
        fun onPurchaseFailure(error: Error)
    }

    interface OnQueryActivePurchasesListener {
        fun onSuccess(activePurchases: List<Purchase>)
        fun onFailure(error: Error)
    }

    interface OnQueryPurchaseHistoryListener {
        fun onSuccess(purchaseHistoryList: List<PurchaseHistoryRecord>)
        fun onFailure(error: Error)
    }

    var onPurchaseListener: OnPurchaseListener? = null

    class Error(val responseCode: Int, val debugMessage: String)

    private val billingClient = BillingClient
        .newBuilder(context)
        .enablePendingPurchases()
        .setListener(this)
        .build()

    fun purchase(activity: Activity, product: SkuDetails) {
        onConnected {
            activity.runOnUiThread {
                billingClient.launchBillingFlow(
                    activity,
                    BillingFlowParams.newBuilder().setSkuDetails(product).build()
                )
            }
        }
    }

    fun queryProducts(listener: OnQueryProductsListener) {
        val skusList = listOf("rsstudy.month")

        queryProductsForType(
            skusList,
            BillingClient.SkuType.SUBS
        ) { billingResult, skuDetailsList ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                val products = skuDetailsList ?: mutableListOf()
                queryProductsForType(
                    skusList,
                    BillingClient.SkuType.INAPP
                ) { billingResult, skuDetailsList ->
                    if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                        products.addAll(skuDetailsList ?: listOf())
                        listener.onSuccess(products)
                    } else {
                        listener.onFailure(
                            Error(billingResult.responseCode, billingResult.debugMessage)
                        )
                    }
                }
            } else {
                listener.onFailure(
                    Error(billingResult.responseCode, billingResult.debugMessage)
                )
            }
        }
    }

    private fun queryProductsForType(
        skusList: List<String>,
        @BillingClient.SkuType type: String,
        listener: SkuDetailsResponseListener
    ) {
        onConnected {
            billingClient.querySkuDetailsAsync(
                SkuDetailsParams.newBuilder().setSkusList(skusList).setType(type).build(),
                listener
            )
        }
    }

    private fun onConnected(block: () -> Unit) {
        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(billingResult: BillingResult) {
                block()
            }

            override fun onBillingServiceDisconnected() {
                // Try to restart the connection on the next request to
                // Google Play by calling the startConnection() method.
            }
        })
    }

    override fun onPurchasesUpdated(
        billingResult: BillingResult,
        purchaseList: MutableList<Purchase>?
    ) {
        when (billingResult.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                if (purchaseList == null) {
                    //to be discussed in the next article
                    onPurchaseListener?.onPurchaseSuccess(null)
                    return
                }

                purchaseList.forEach(::processPurchase) //to be declared below
            }
            else -> {
                //error occurred or user canceled
                onPurchaseListener?.onPurchaseFailure(
                    BillingClientHelper.Error(
                        billingResult.responseCode,
                        billingResult.debugMessage
                    )
                )
            }
        }
    }

    private fun acknowledgePurchase(
        purchase: Purchase,
        callback: AcknowledgePurchaseResponseListener
    ) {
        onConnected {
            billingClient.acknowledgePurchase(
                AcknowledgePurchaseParams.newBuilder().setPurchaseToken(purchase.purchaseToken)
                    .build(),
                callback::onAcknowledgePurchaseResponse
            )
        }
    }

    private fun consumePurchase(purchase: Purchase, callback: ConsumeResponseListener) {
        onConnected {
            billingClient.consumeAsync(
                ConsumeParams.newBuilder().setPurchaseToken(purchase.purchaseToken).build()
            ) { billingResult, purchaseToken ->
                callback.onConsumeResponse(billingResult, purchaseToken)
            }
        }
    }

    private fun processPurchase(purchase: Purchase) {
        if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
            onPurchaseListener?.onPurchaseSuccess(purchase)

            if (purchase.skus.firstOrNull() == "coin_pack_large") {
                //consuming our only consumable product
                consumePurchase(purchase) { billingResult, purchaseToken ->
                    if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
                        //implement retry logic or try to consume again in onResume()
                    }
                }
            } else if (!purchase.isAcknowledged) {
                acknowledgePurchase(purchase) { billingResult ->
                    if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) {
                        //implement retry logic or try to acknowledge again in onResume()
                    }
                }
            }
        }
    }

    private fun queryActivePurchasesForType(
        @BillingClient.SkuType type: String,
        listener: PurchasesResponseListener
    ) {
        onConnected {
            billingClient.queryPurchasesAsync(type, listener)
        }
    }

    fun queryActivePurchases(listener: OnQueryActivePurchasesListener) {
        queryActivePurchasesForType(
            BillingClient.SkuType.SUBS
        ) { billingResult, activeSubsList ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                queryActivePurchasesForType(
                    BillingClient.SkuType.INAPP
                ) { billingResult, nonConsumableProductsList ->
                    if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                        listener.onSuccess(
                            activeSubsList.apply { addAll(nonConsumableProductsList) }
                        )
                    } else {
                        listener.onFailure(
                            Error(billingResult.responseCode, billingResult.debugMessage)
                        )
                    }
                }
            } else {
                listener.onFailure(
                    Error(billingResult.responseCode, billingResult.debugMessage)
                )
            }
        }
    }

    private fun queryPurchasesHistoryForType(
        @BillingClient.SkuType type: String,
        listener: PurchaseHistoryResponseListener
    ) {
        onConnected {
            billingClient.queryPurchaseHistoryAsync(type, listener)
        }
    }

    fun queryPurchaseHistory(listener: OnQueryPurchaseHistoryListener) {
        queryPurchasesHistoryForType(
            BillingClient.SkuType.SUBS
        ) { billingResult, subsHistoryList ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                queryPurchasesHistoryForType(
                    BillingClient.SkuType.INAPP
                ) { billingResult, inappHistoryList ->
                    if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                        listener.onSuccess(
                            subsHistoryList?.apply { addAll(inappHistoryList.orEmpty()) }.orEmpty()
                        )
                    } else {
                        listener.onFailure(
                            Error(billingResult.responseCode, billingResult.debugMessage)
                        )
                    }
                }
            } else {
                listener.onFailure(
                    Error(billingResult.responseCode, billingResult.debugMessage)
                )
            }
        }
    }

    fun querySyncedActivePurchases(listener: OnQueryActivePurchasesListener) {
        queryPurchaseHistory(object : OnQueryPurchaseHistoryListener {
            override fun onSuccess(purchaseHistoryList: List<PurchaseHistoryRecord>) {
                queryActivePurchases(listener)
            }

            override fun onFailure(error: Error) {
                listener.onFailure(error)
            }
        })
    }

    fun changeSubscription(
        activity: Activity,
        newSub: SkuDetails,
        updateParams: BillingFlowParams.SubscriptionUpdateParams
    ) {
        onConnected {
            activity.runOnUiThread {
                billingClient.launchBillingFlow(
                    activity,
                    BillingFlowParams.newBuilder().setSkuDetails(newSub)
                        .setSubscriptionUpdateParams(updateParams).build()
                )
            }
        }
    }
}